File: src/dnfw/container/section.py
"""One ELE3 section: the 8-byte header in front of a compressed stream, and
the question of whether a section is compressed at all.

    [u32 stream length BE][u32 sum of the stream bytes BE][stream ...]

Not every section has this. On both Digitones `meta`, `updater` and `boot` --
and DN1's `blob` -- are stored raw, and the only reliable test is whether the
bytes depack. So a section is modelled as its **stored** bytes, exactly as they
sit in the container, and compression is something we discover rather than
assume. `dnfw build` must never recompress a section it could not depack.

**A compressed section is zero-padded to a multiple of four bytes.** The header
still gives the exact stream length and the sum of the stream alone; the
padding sits after the stream and is counted only in the section table's
length. Every compressed section of Digitone II OS 1.10D, 1.10E and 1.11 --
eleven of them -- follows this, with 0 to 3 bytes of padding each. Images
built without it boot through the normal update path and stall in recovery,
so it is written here rather than left to chance. See `docs/flashing.md`.
"""

from dataclasses import dataclass

from ..codec import aplib, aplibpack

HEADER = 8
STORED_ALIGN = 4  # compressed sections are zero-padded to this


@dataclass(frozen=True)
class Section:
    """A section as the container holds it."""

    id: int
    dest: int
    stored: bytes

    @property
    def declared_length(self) -> int:
        """Stream length from the header. Meaningless on a raw section."""
        return int.from_bytes(self.stored[0:4], "big")

    @property
    def declared_sum(self) -> int:
        return int.from_bytes(self.stored[4:8], "big")

    @property
    def computed_sum(self) -> int:
        """Sum of the declared stream bytes, which is what the header claims."""
        end = min(HEADER + self.declared_length, len(self.stored))
        return sum(self.stored[HEADER:end]) & 0xFFFFFFFF

    @property
    def stream(self) -> bytes:
        """The compressed stream the header describes, header excluded.
        Meaningless on a raw section."""
        return self.stored[HEADER : HEADER + self.declared_length]

    @property
    def sum_ok(self) -> bool:
        return self.declared_sum == self.computed_sum

    def unpack(self) -> bytes | None:
        """The section content, or None if these bytes are not an aPLib stream.

        Truncation is tolerated: a raw section can decode a few plausible bytes
        before running out, so the caller distinguishes raw from compressed by
        None rather than by trusting a length field it has no reason to.
        """
        if len(self.stored) <= HEADER:
            return None
        try:
            out = aplib.depack(self.stored[HEADER:], allow_truncated=True)
        except aplib.DepackError:
            return None
        return out or None


def compress(section_id: int, dest: int, content: bytes) -> Section:
    """Build a compressed section from raw content: header, stream, padding."""
    stream = aplibpack.pack(content)
    header = len(stream).to_bytes(4, "big") + (sum(stream) & 0xFFFFFFFF).to_bytes(4, "big")
    padding = bytes(-(HEADER + len(stream)) % STORED_ALIGN)
    return Section(id=section_id, dest=dest, stored=header + stream + padding)


def store_raw(section_id: int, dest: int, content: bytes) -> Section:
    """Build a section that is stored verbatim, for the sections that are."""
    return Section(id=section_id, dest=dest, stored=content)

File: site/js/container/section.js
/**
 * One ELE3 section: the 8-byte header in front of a compressed stream, and the
 * question of whether a section is compressed at all.
 *
 *     [u32 stream length BE][u32 sum of the stream bytes BE][stream ...]
 *
 * Port of `src/dnfw/container/section.py`. Not every section has this header.
 * On both Digitones `meta`, `updater` and `boot` are stored raw, and the only
 * reliable test is whether the bytes depack -- so a section is modelled as its
 * **stored** bytes, exactly as they sit in the container, and compression is
 * discovered rather than assumed. A rebuild must never recompress a section it
 * could not depack: getting that backwards produces a file that passes every
 * checksum and bricks the instrument.
 *
 * **A compressed section is zero-padded to a multiple of four bytes.** Images
 * built without that padding boot through the normal update path and stall in
 * recovery, so it is written here rather than left to chance.
 */

import { u32 } from "../bytes.js";
import { DepackError, depack } from "../aplib.js";
import { pack } from "../aplibpack.js";

export const HEADER = 8;
export const STORED_ALIGN = 4;

export class Section {
  constructor(id, dest, stored) {
    this.id = id;
    this.dest = dest;
    this.stored = stored;
    this._content = undefined;   // unpack() is memoised; it is not cheap
  }

  /** Stream length from the header. Meaningless on a raw section. */
  get declaredLength() {
    return u32(this.stored, 0);
  }

  /** The compressed stream the header describes, header excluded. */
  get stream() {
    return this.stored.subarray(HEADER, HEADER + this.declaredLength);
  }

  /**
   * What a *raw* section actually loads at `dest`, header excluded.
   *
   * The test is the declared sum, not the section id, because deciding it from
   * the id would be a guess and this is a check: a real header whose stream is
   * stored rather than packed has sum 0 -- there is no stream to sum -- while a
   * section that is nothing but payload has arbitrary bytes there.
   */
  get rawPayload() {
    if (this.stored.length > HEADER && u32(this.stored, 4) === 0) {
      return this.stored.subarray(HEADER);
    }
    return this.stored;
  }

  /**
   * The section content, or `null` if these bytes are not an aPLib stream.
   *
   * Truncation is tolerated: a raw section can decode a few plausible bytes
   * before running out, so the caller distinguishes raw from compressed by
   * `null` rather than by trusting a length field it has no reason to.
   */
  unpack() {
    if (this._content !== undefined) return this._content;
    this._content = null;
    if (this.stored.length > HEADER) {
      try {
        const out = depack(this.stored.subarray(HEADER), { allowTruncated: true });
        if (out.length) this._content = out;
      } catch (error) {
        if (!(error instanceof DepackError)) throw error;
      }
    }
    return this._content;
  }
}

/** Build a compressed section from raw content: header, stream, padding. */
export function compress(id, dest, content) {
  const stream = pack(content);
  let sum = 0;
  for (let i = 0; i < stream.length; i++) sum = (sum + stream[i]) >>> 0;
  const padding = (-(HEADER + stream.length) % STORED_ALIGN + STORED_ALIGN) % STORED_ALIGN;

  const stored = new Uint8Array(HEADER + stream.length + padding);
  const view = new DataView(stored.buffer);
  view.setUint32(0, stream.length, false);
  view.setUint32(4, sum, false);
  stored.set(stream, HEADER);
  return new Section(id, dest, stored);
}

/** Build a section stored verbatim, for the sections that are. */
export function storeRaw(id, dest, content) {
  return new Section(id, dest, content);
}
